// @livepeer/core-react/utils/dist/livepeer-core-react-utils.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var client_dist_livepeerCoreReactClient = require('../../dist/client-b79b683b.cjs.prod.js');
var useInternalQuery = require('../../dist/useInternalQuery-c049f707.cjs.prod.js');
require('@livepeer/core/client');
require('@tanstack/query-async-storage-persister');
require('@tanstack/react-query');
require('@tanstack/react-query-persist-client');
require('@livepeer/core');
require('react');
require('../../dist/LivepeerConfig-f7b8e2cd.cjs.prod.js');
require('react/jsx-runtime');



exports.deserialize = client_dist_livepeerCoreReactClient.deserialize;
exports.serialize = client_dist_livepeerCoreReactClient.serialize;
exports.prefetchQuery = useInternalQuery.prefetchQuery;
exports.useInternalMutation = useInternalQuery.useInternalMutation;
exports.useInternalMutationKeys = useInternalQuery.useInternalMutationKeys;
exports.useInternalQuery = useInternalQuery.useInternalQuery;
exports.usePickQueryKeys = useInternalQuery.usePickQueryKeys;
