// @livepeer/core-react/utils/dist/livepeer-core-react-utils.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var client_dist_livepeerCoreReactClient = require('../../dist/client-90c13eab.cjs.dev.js');
var useInternalQuery = require('../../dist/useInternalQuery-34f85f2d.cjs.dev.js');
require('@livepeer/core/client');
require('@tanstack/query-async-storage-persister');
require('@tanstack/react-query');
require('@tanstack/react-query-persist-client');
require('@livepeer/core');
require('react');
require('../../dist/LivepeerConfig-5a11a3d5.cjs.dev.js');
require('react/jsx-runtime');



exports.deserialize = client_dist_livepeerCoreReactClient.deserialize;
exports.serialize = client_dist_livepeerCoreReactClient.serialize;
exports.prefetchQuery = useInternalQuery.prefetchQuery;
exports.useInternalMutation = useInternalQuery.useInternalMutation;
exports.useInternalMutationKeys = useInternalQuery.useInternalMutationKeys;
exports.useInternalQuery = useInternalQuery.useInternalQuery;
exports.usePickQueryKeys = useInternalQuery.usePickQueryKeys;
