// @livepeer/react/hooks/dist/livepeer-react-hooks.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var hooks = require('@livepeer/core-react/hooks');



Object.defineProperty(exports, 'prefetchAsset', {
  enumerable: true,
  get: function () { return hooks.prefetchAsset; }
});
Object.defineProperty(exports, 'prefetchAssetMetrics', {
  enumerable: true,
  get: function () { return hooks.prefetchAssetMetrics; }
});
Object.defineProperty(exports, 'prefetchPlaybackInfo', {
  enumerable: true,
  get: function () { return hooks.prefetchPlaybackInfo; }
});
Object.defineProperty(exports, 'prefetchPlayer', {
  enumerable: true,
  get: function () { return hooks.prefetchPlayer; }
});
Object.defineProperty(exports, 'prefetchStream', {
  enumerable: true,
  get: function () { return hooks.prefetchStream; }
});
Object.defineProperty(exports, 'prefetchStreamSession', {
  enumerable: true,
  get: function () { return hooks.prefetchStreamSession; }
});
Object.defineProperty(exports, 'prefetchStreamSessions', {
  enumerable: true,
  get: function () { return hooks.prefetchStreamSessions; }
});
Object.defineProperty(exports, 'useAsset', {
  enumerable: true,
  get: function () { return hooks.useAsset; }
});
Object.defineProperty(exports, 'useAssetMetrics', {
  enumerable: true,
  get: function () { return hooks.useAssetMetrics; }
});
Object.defineProperty(exports, 'useCreateAsset', {
  enumerable: true,
  get: function () { return hooks.useCreateAsset; }
});
Object.defineProperty(exports, 'useCreateClip', {
  enumerable: true,
  get: function () { return hooks.useCreateClip; }
});
Object.defineProperty(exports, 'useCreateStream', {
  enumerable: true,
  get: function () { return hooks.useCreateStream; }
});
Object.defineProperty(exports, 'useLivepeerProvider', {
  enumerable: true,
  get: function () { return hooks.useLivepeerProvider; }
});
Object.defineProperty(exports, 'usePlaybackInfo', {
  enumerable: true,
  get: function () { return hooks.usePlaybackInfo; }
});
Object.defineProperty(exports, 'useStream', {
  enumerable: true,
  get: function () { return hooks.useStream; }
});
Object.defineProperty(exports, 'useStreamSession', {
  enumerable: true,
  get: function () { return hooks.useStreamSession; }
});
Object.defineProperty(exports, 'useStreamSessions', {
  enumerable: true,
  get: function () { return hooks.useStreamSessions; }
});
Object.defineProperty(exports, 'useUpdateAsset', {
  enumerable: true,
  get: function () { return hooks.useUpdateAsset; }
});
Object.defineProperty(exports, 'useUpdateStream', {
  enumerable: true,
  get: function () { return hooks.useUpdateStream; }
});
