// @livepeer/react/dist/LivepeerConfig-dc30fefa.cjs.dev.js
'use strict';

var context = require('@livepeer/core-react/context');
require('react');
var useTheme = require('./useTheme-6b039219.cjs.dev.js');
var jsxRuntime = require('react/jsx-runtime');

var ThemeProvider = function ThemeProvider(_ref) {
  var theme = _ref.theme,
    children = _ref.children;
  return /*#__PURE__*/jsxRuntime.jsx(useTheme.ThemeContext.Provider, {
    value: theme !== null && theme !== void 0 ? theme : null,
    children: children
  });
};

var Context = context.Context;
var useClient = context.useClient;

// Extends the core livepeer config to provide theming

function LivepeerConfig(_ref) {
  var children = _ref.children,
    client = _ref.client,
    theme = _ref.theme,
    dehydratedState = _ref.dehydratedState;
  return /*#__PURE__*/jsxRuntime.jsx(context.LivepeerConfig, {
    client: client,
    dehydratedState: dehydratedState,
    children: /*#__PURE__*/jsxRuntime.jsx(ThemeProvider, {
      theme: theme,
      children: children
    })
  });
}

exports.Context = Context;
exports.LivepeerConfig = LivepeerConfig;
exports.ThemeProvider = ThemeProvider;
exports.useClient = useClient;
