// @livepeer/react/dist/livepeer-react.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var coreReact = require('@livepeer/core-react');
var styling = require('@livepeer/core-web/media/browser/styling');
var studio = require('@livepeer/core-web/providers/studio');
var client_dist_livepeerReactClient = require('../client/dist/livepeer-react-client.cjs.prod.js');
var client = require('@livepeer/core-react/client');
var Player = require('./Player-f3160066.cjs.prod.js');
var LivepeerConfig = require('./LivepeerConfig-a10b7e18.cjs.prod.js');
var useTheme = require('./useTheme-b35069a8.cjs.prod.js');
var hooks = require('@livepeer/core-react/hooks');
require('./objectSpread2-15199050.cjs.prod.js');
require('core-js/modules/es.object.assign.js');
require('@livepeer/core-react/components');
require('react');
require('core-js/modules/es.object.to-string.js');
require('core-js/modules/es.regexp.to-string.js');
require('@livepeer/core-web');
require('@livepeer/core-web/media/browser/webrtc');
require('react/jsx-runtime');
require('@radix-ui/react-dialog');
require('core-js/modules/es.array.filter.js');
require('core-js/modules/es.array.concat.js');
require('core-js/modules/es.array.slice.js');
require('core-js/modules/es.array.map.js');
require('@radix-ui/react-select');
require('core-js/modules/es.array.iterator.js');
require('core-js/modules/es.string.iterator.js');
require('core-js/modules/web.dom-collections.iterator.js');
require('core-js/modules/web.url.js');
require('core-js/modules/web.url-search-params.js');
require('@livepeer/core-web/media/browser');
require('@livepeer/core-web/media/browser/hls');
require('core-js/modules/es.regexp.exec.js');
require('core-js/modules/es.string.match.js');
require('core-js/modules/es.promise.js');
require('@livepeer/core-react/context');
require('@livepeer/core-web/media');
require('zustand');



Object.defineProperty(exports, 'defaultTheme', {
  enumerable: true,
  get: function () { return styling.defaultTheme; }
});
Object.defineProperty(exports, 'getCssText', {
  enumerable: true,
  get: function () { return styling.getCssText; }
});
Object.defineProperty(exports, 'styling', {
  enumerable: true,
  get: function () { return styling.styling; }
});
Object.defineProperty(exports, 'studioProvider', {
  enumerable: true,
  get: function () { return studio.studioProvider; }
});
exports.createReactClient = client_dist_livepeerReactClient.createReactClient;
Object.defineProperty(exports, 'defaultQueryClient', {
  enumerable: true,
  get: function () { return client.defaultQueryClient; }
});
exports.AudioSourceSelect = Player.AudioSourceSelect;
exports.AudioToggle = Player.AudioToggle;
exports.BaseSlider = Player.BaseSlider;
exports.Broadcast = Player.Broadcast;
exports.BroadcastSettings = Player.BroadcastSettings;
exports.ControlsContainer = Player.ControlsContainer;
exports.FullscreenButton = Player.FullscreenButton;
exports.PictureInPictureButton = Player.PictureInPictureButton;
exports.PlayButton = Player.PlayButton;
exports.Player = Player.Player;
exports.Poster = Player.Poster;
exports.Progress = Player.Progress;
exports.Screenshare = Player.Screenshare;
exports.TimeDisplay = Player.TimeDisplay;
exports.Title = Player.Title;
exports.VideoSourceSelect = Player.VideoSourceSelect;
exports.VideoToggle = Player.VideoToggle;
exports.Volume = Player.Volume;
exports.Context = LivepeerConfig.Context;
exports.LivepeerConfig = LivepeerConfig.LivepeerConfig;
exports.ThemeProvider = LivepeerConfig.ThemeProvider;
exports.useClient = LivepeerConfig.useClient;
exports.MediaControllerProvider = useTheme.MediaControllerProvider;
exports.useMediaController = useTheme.useMediaController;
exports.useTheme = useTheme.useTheme;
Object.defineProperty(exports, 'prefetchAsset', {
  enumerable: true,
  get: function () { return hooks.prefetchAsset; }
});
Object.defineProperty(exports, 'prefetchAssetMetrics', {
  enumerable: true,
  get: function () { return hooks.prefetchAssetMetrics; }
});
Object.defineProperty(exports, 'prefetchPlaybackInfo', {
  enumerable: true,
  get: function () { return hooks.prefetchPlaybackInfo; }
});
Object.defineProperty(exports, 'prefetchPlayer', {
  enumerable: true,
  get: function () { return hooks.prefetchPlayer; }
});
Object.defineProperty(exports, 'prefetchStream', {
  enumerable: true,
  get: function () { return hooks.prefetchStream; }
});
Object.defineProperty(exports, 'prefetchStreamSession', {
  enumerable: true,
  get: function () { return hooks.prefetchStreamSession; }
});
Object.defineProperty(exports, 'prefetchStreamSessions', {
  enumerable: true,
  get: function () { return hooks.prefetchStreamSessions; }
});
Object.defineProperty(exports, 'useAsset', {
  enumerable: true,
  get: function () { return hooks.useAsset; }
});
Object.defineProperty(exports, 'useAssetMetrics', {
  enumerable: true,
  get: function () { return hooks.useAssetMetrics; }
});
Object.defineProperty(exports, 'useCreateAsset', {
  enumerable: true,
  get: function () { return hooks.useCreateAsset; }
});
Object.defineProperty(exports, 'useCreateClip', {
  enumerable: true,
  get: function () { return hooks.useCreateClip; }
});
Object.defineProperty(exports, 'useCreateStream', {
  enumerable: true,
  get: function () { return hooks.useCreateStream; }
});
Object.defineProperty(exports, 'useLivepeerProvider', {
  enumerable: true,
  get: function () { return hooks.useLivepeerProvider; }
});
Object.defineProperty(exports, 'usePlaybackInfo', {
  enumerable: true,
  get: function () { return hooks.usePlaybackInfo; }
});
Object.defineProperty(exports, 'useStream', {
  enumerable: true,
  get: function () { return hooks.useStream; }
});
Object.defineProperty(exports, 'useStreamSession', {
  enumerable: true,
  get: function () { return hooks.useStreamSession; }
});
Object.defineProperty(exports, 'useStreamSessions', {
  enumerable: true,
  get: function () { return hooks.useStreamSessions; }
});
Object.defineProperty(exports, 'useUpdateAsset', {
  enumerable: true,
  get: function () { return hooks.useUpdateAsset; }
});
Object.defineProperty(exports, 'useUpdateStream', {
  enumerable: true,
  get: function () { return hooks.useUpdateStream; }
});
Object.keys(coreReact).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return coreReact[k]; }
  });
});
